test-data/unit/check-typeis.test

```diff
@@ -0,0 +1,729 @@
+
+class A: ...
+def f(x: List[Any]) -> TypeIs[List[str]]: ...
+def g(x: object) -> None: ...
+
+def test(x: List[Any]) -> None:
+    if not(f(x) or isinstance(x, A)):
+        return
+    g(reveal_type(x))  # N: Revealed type is "Union[builtins.list[builtins.str], __main__.<subclass of "list" and "A">]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsMultipleCondition-xfail]
+from typing_extensions import TypeIs
+from typing import Any, List
+
+class Foo: ...
+class Bar: ...
+
+def is_foo(item: object) -> TypeIs[Foo]:
+    return isinstance(item, Foo)
+
+def is_bar(item: object) -> TypeIs[Bar]:
+    return isinstance(item, Bar)
+
+def foobar(x: object):
+    if not isinstance(x, Foo) or not isinstance(x, Bar):
+        return
+    reveal_type(x)  # N: Revealed type is "__main__.<subclass of "Foo" and "Bar">"
+
+def foobar_typeguard(x: object):
+    if not is_foo(x) or not is_bar(x):
+        return
+    reveal_type(x)  # N: Revealed type is "__main__.<subclass of "Foo" and "Bar">"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsAsFunctionArgAsBoolSubtype]
+from typing import Callable
+from typing_extensions import TypeIs
+
+def accepts_bool(f: Callable[[object], bool]): pass
+
+def with_bool_typeguard(o: object) -> TypeIs[bool]: pass
+def with_str_typeguard(o: object) -> TypeIs[str]: pass
+def with_bool(o: object) -> bool: pass
+
+accepts_bool(with_bool_typeguard)
+accepts_bool(with_str_typeguard)
+accepts_bool(with_bool)
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsAsFunctionArg]
+from typing import Callable
+from typing_extensions import TypeIs
+
+def accepts_typeguard(f: Callable[[object], TypeIs[bool]]): pass
+def different_typeguard(f: Callable[[object], TypeIs[str]]): pass
+
+def with_typeguard(o: object) -> TypeIs[bool]: pass
+def with_bool(o: object) -> bool: pass
+
+accepts_typeguard(with_typeguard)
+accepts_typeguard(with_bool)  # E: Argument 1 to "accepts_typeguard" has incompatible type "Callable[[object], bool]"; expected "Callable[[object], TypeIs[bool]]"
+
+different_typeguard(with_typeguard)  # E: Argument 1 to "different_typeguard" has incompatible type "Callable[[object], TypeIs[bool]]"; expected "Callable[[object], TypeIs[str]]"
+different_typeguard(with_bool)  # E: Argument 1 to "different_typeguard" has incompatible type "Callable[[object], bool]"; expected "Callable[[object], TypeIs[str]]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsAsGenericFunctionArg]
+from typing import Callable, TypeVar
+from typing_extensions import TypeIs
+
+T = TypeVar('T')
+
+def accepts_typeguard(f: Callable[[object], TypeIs[T]]): pass
+
+def with_bool_typeguard(o: object) -> TypeIs[bool]: pass
+def with_str_typeguard(o: object) -> TypeIs[str]: pass
+def with_bool(o: object) -> bool: pass
+
+accepts_typeguard(with_bool_typeguard)
+accepts_typeguard(with_str_typeguard)
+accepts_typeguard(with_bool)  # E: Argument 1 to "accepts_typeguard" has incompatible type "Callable[[object], bool]"; expected "Callable[[object], TypeIs[bool]]"
```

Comment:
Why isn't it this? `T` in `accepts_typeguard` should be unbound, i.e. `Any`. _Unless I'm missing something._
```suggestion
accepts_typeguard(with_bool)  # E: Argument 1 to "accepts_typeguard" has incompatible type "Callable[[object], bool]"; expected "Callable[[object], TypeIs[Any]]"
```

Reply:
hm good question, might be a bad type context somewhere. I copied these tests from TypeGuard, which I think works the same.

Reply:
Yes, this also fails with TypeGuard: https://github.com/python/mypy/blob/790e8a73d8671a41cae419b4ea07579bfb2bc292/test-data/unit/check-typeguard.test#L507

It seems to be due to the code in `constraints.py`, which already looked somewhat questionable. I found a possible fix, but now it's inferring `Never` instead.

Reply:
I think `Never` might actually be correct.? At least in the mypy logic. The TypeVar is only used once, so mypy isn't able to solve the constraint and thus infers `Never`. Something similar happens with empty dicts sometimes.

It's not ideal, but users just have to know that they can use `Any` there instead.

_Anyway, I don't think this should be blocking this PR. If necessary, this can always be improved later._